// @cucumber/cucumber/lib/plugin/plugin_manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginManager = void 0;
class PluginManager {
    pluginFns;
    handlers = { message: [] };
    cleanupFns = [];
    constructor(pluginFns) {
        this.pluginFns = pluginFns;
    }
    async register(event, handler) {
        this.handlers[event].push(handler);
    }
    async init(logger, configuration, environment) {
        for (const pluginFn of this.pluginFns) {
            const cleanupFn = await pluginFn({
                on: this.register.bind(this),
                logger,
                configuration,
                environment,
            });
            if (cleanupFn) {
                this.cleanupFns.push(cleanupFn);
            }
        }
    }
    emit(event, value) {
        this.handlers[event].forEach((handler) => handler(value));
    }
    async cleanup() {
        for (const cleanupFn of this.cleanupFns) {
            await cleanupFn();
        }
    }
}
exports.PluginManager = PluginManager;
